A simple FixedPoint executioner for testing iterative solvers with tagging

YaqiWang opened the pull request:
Refer to #9669.

I like the tagging system! Its combination with the new design of our executioner makes we can do really fancy solving. This PR mimics what I will do in Rattlesnake with minimum coding. Although I put `FixedPoint` under test folder, it could be generally usable. The test is basically a pseudo transient but without using `Transient` executioner.

One small thing noticed is that how do we handle tagging with non-homogeneous Dirichlet boundary conditions?

YaqiWang committed:
a simple FixedPoint executioner for testing iterative solvers with tagging #9669

YaqiWang commented:
This is the alternative approach of coupled tagged vectors I mentioned in #13029.

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/302786/) on 43a97cf wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/13033/site)

This comment will be updated on new commits.

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/302844/) on 69b7cd1 : invalidated by @YaqiWang

retest

YaqiWang commented:
This is ready for review. @fdkong @lindsayad @friedmud @permcody care to do it?

fdkong commented:
> One small thing noticed is that how do we handle tagging with non-homogeneous Dirichlet boundary conditions?

Tagging should just work for any boundary conditions if you already marked them correctly

YaqiWang commented:
I am guessing because I did not let the Dirichlet BC tag the residual vector, the residual on boundary nodes are not set to zero. Then when I add the residual vectors together, Dirichlet BC are not imposed properly. That is my guess. This can be easily produced by change the BCs in the test to DirichletBC.

YaqiWang commented:
Be a little clearer: if you set `BCs/left/type=DirichletBC BCs/right/type=DirichletBC` and do the calculations with and without `Kernels/pseudo_time/vector_tags`, you will get different answers although they should be the same.

friedmud commented:
Dirichlet conditions do not "add" into a residual vector... they "set".  The only way to get the action of a Dirichlet condition in a vector tag is to tag the Dirichlet condition in that tag.

YaqiWang commented:
Right. This does fix the issue:
```
../../../moose_test-opt -i 2d_diffusion_fixed_point.i BCs/left/type=DirichletBC BCs/left/value=0 BCs/right/type=DirichletBC BCs/left/extra_vector_tags=previous BCs/right/extra_vector_tags=previous
```
and
```
../../../moose_test-opt -i 2d_diffusion_fixed_point.i BCs/left/type=DirichletBC BCs/left/value=0 BCs/right/type=DirichletBC BCs/left/extra_vector_tags=previous BCs/right/extra_vector_tags=previous  Kernels/pseudo_time/vector_tags=nontime
```
produce the same answer.

YaqiWang commented:
Any more comments?

fdkong commented on test/src/problems/FixedPointProblem.C:
```diff
@@ -0,0 +1,58 @@
+    if (tag.second != _tag_id)
+      _fe_vector_tags.insert(tag.second);
+
+  computeResidualInternal(soln, residual, _fe_vector_tags);
```
It is way better to use `computeResidualTags` since it is easier to understand, and much more clear.

`computeResidualInternal` is something I really want to delete, but I can not delete because so many people are using this.

YaqiWang replied:
Will change this.

YaqiWang replied:
Actually this function is mimic `FEProblemBase::computeResidual`, which is also calling `computeResidualInternal`.

YaqiWang commented on test/src/problems/FixedPointProblem.C:
```diff
@@ -0,0 +1,58 @@
+{
+  _residual_previous.close();
+  FEProblem::computeResidual(soln, residual);
+  residual += _residual_previous;
```
Nope, some of the kernels are only tag to `_residual_previous`, which is the way how we separate the residuals.

YaqiWang commented on test/include/problems/FixedPointProblem.h:
```diff
@@ -0,0 +1,39 @@
+/**
+ * FEProblem derived class for lagging a tagged residual.
+ */
+class FixedPointProblem : public FEProblem
```
Yes to the first. No to the second. We cannot derive this from `FEProblemBase` because both` FEProblem` and `EigenProblem` are not going to have the code in this `FixedPointProblem`. To do that, we will have to derive `FEProblem` and `EigenProblem` separately. We can use an interface to remove the code duplication. But it could be desired to merge `EigenProblem` into `FEProblem` if possible.

YaqiWang replied:
This is actually quite an issue to me: @fdkong can we merge `EigenProblem` into `FEProblem`?

fdkong commented:
It looks fine for me. But @permcody and @friedmud have to decide if more top-level iterative algorithms are acceptable in moose??? If we are going in this direction, what I worry about is that with the tagging system users can construct arbitrarily  complicated  algorithms on top of FEProblem-Solve. It might/may become hard  to maintain.

Usually, I like to construct the algorithm  under the bottom of SNES unless the algorithms have to be an outer loop. Not other way around.

YaqiWang commented on test/src/executioners/FixedPoint.C:
```diff
@@ -0,0 +1,95 @@
+      return false;
+    }
+
+    _fp_problem.computeFullResidual(*_nl.currentSolution(), _nl.RHS());
```
No. I know what you mean. But this question is not for this PR because `FixedPointProblem::solve` will do a full Newton solve here.

YaqiWang replied:
This full residual is actually the same as the initial residual of the next fixed point iteration. But I have no way to retrieve that for convergence check needed here.

YaqiWang commented on test/src/problems/FixedPointProblem.C:
```diff
@@ -0,0 +1,58 @@
+FixedPointProblem::computeFullResidual(const NumericVector<Number> & soln,
+                                       NumericVector<Number> & residual)
+{
+  _residual_previous.close();
```
Otherwise I got an assertion error in debug mode. Initially I thought I do not need this. I am lazy to figure out where the close is missing. On the other hand, if `close` has a quick return based on `closed()`, the performance is not hurt.

fdkong replied:
There is no way that  `close` can quickly return without a global communication.  It is definitely impossible. 

Suppose there are three mpi processes:

mpi 0: closed_flag: true

mpi 1: closed_flag: false (triggered by any `MatSetValues`)

mpi 2: closed_flag: true

Only everybody has true, and we can not return. How `mpi` 0 an `mpi` 2 know the flag should be false without the communication.

YaqiWang replied:
Although I am not fully convinced about that closing a closed vector is costly, I agree that we should remove unnecessary close calls. This call is immediately after `FEProblem::solve`, and we would think all vectors have been properly closed. But it is not here. It is pretty hard to figure out what is missing in the `solve`.

YaqiWang replied:
I will spend a little more time on why the tagged vector is not closed properly.

fdkong commented on test/src/executioners/FixedPoint.C:
```diff
@@ -0,0 +1,95 @@
+    _console << "Fixed point iteration " << it << std::endl;
+
+    Real residual_norm_previous = residual_norm;
+    if (!_inner_solve->solve())
```
Will this involve a Newton solve?  And line 65 has an extra residual evaluation?

I am asking because I am curious about what is the benefit of the source iteration compared with GMRES. It would be even more interesting to have comparisons for the same problem

fdkong replied:
Never mind. It looks useful for the nonlinear eigenvalue system

friedmud commented:
I guess I’ll chime in here and say that I’m not super excited about adding
something so complicated to MOOSE directly.   I worry that many won’t
hnderstand what it does and will use it incorrectly and we might have a
hard time maintaining it.

Let’s have a discussion about it when I get back.  I may be convinced
otherwise...

On Tue, Mar 12, 2019 at 4:43 PM Fande Kong <notifications@github.com> wrote:

> *@fdkong* commented on this pull request.
> ------------------------------
>
> In test/src/executioners/FixedPoint.C
> <https://github.com/idaholab/moose/pull/13033#discussion_r264897011>:
>
> > +bool
> +FixedPoint::solve()
> +{
> +  // evaluate kernels on previous solution
> +  _fp_problem.computeFullResidual(*_nl.currentSolution(), _nl.RHS());
> +  Real initial_residual_norm = _nl.RHS().l2_norm();
> +  Real residual_norm = initial_residual_norm;
> +  _console << "Fixed point initial residual norm " << residual_norm << std::endl;
> +
> +  unsigned int it = 0;
> +  for (; it < _fp_max_its; ++it)
> +  {
> +    _console << "Fixed point iteration " << it << std::endl;
> +
> +    Real residual_norm_previous = residual_norm;
> +    if (!_inner_solve->solve())
>
> Never mind. It looks useful for the nonlinear eigenvalue system
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/idaholab/moose/pull/13033#discussion_r264897011>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AA1JMXBTnoKiF5Dp8Lfxm8zmQ3OnzVLdks5vWB-DgaJpZM4bnffI>
> .
>
-- 
Sent from my iPhone

YaqiWang commented:
Just in case you did not notice. This PR is all in `test`. It is a kind of another way of testing the tagging system. This test provides a protection for the capability that Rattlesnake is going to use. I agree this is a little counter-intuitive because it took me a while to think of this, i.e. saving the partial residual but not the previous solution. No hurry to get this in.

permcody commented:
It's more costly than you think. While the actual communication time isn't
much, it also creates a syncronization point.

On Wed, Mar 13, 2019 at 8:53 AM Yaqi <notifications@github.com> wrote:

> *@YaqiWang* commented on this pull request.
> ------------------------------
>
> In test/src/problems/FixedPointProblem.C
> <https://github.com/idaholab/moose/pull/13033#discussion_r265168086>:
>
> > +  _fe_vector_tags.clear();
> +
> +  for (auto & tag : tags)
> +    if (tag.second != _tag_id)
> +      _fe_vector_tags.insert(tag.second);
> +
> +  computeResidualInternal(soln, residual, _fe_vector_tags);
> +
> +  residual += _residual_previous;
> +}
> +
> +void
> +FixedPointProblem::computeFullResidual(const NumericVector<Number> & soln,
> +                                       NumericVector<Number> & residual)
> +{
> +  _residual_previous.close();
>
> Although I am not fully convinced about that closing a closed vector is
> costly, I agree that we should remove unnecessary close calls. This call is
> immediately after FEProblem::solve, and we would think all vectors have
> been properly closed. But it is not here. It is pretty hard to figure out
> what is missing in the solve.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/idaholab/moose/pull/13033#discussion_r265168086>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AC5XIHWOB0iaX8OS_kTqE9b1LKnFVZt1ks5vWRDogaJpZM4bnffI>
> .
>

YaqiWang commented:
Found a bug. The fix allows us to remove the close statement. I will add another test to demonstrate the application of the fixed point iteration for hard-solving nonlinear problems.

YaqiWang committed:
add another test shows fixed point iteration is less sensitive to the initial guess #9669

YaqiWang commented:
Added another test per discussion with @friedmud. The test is nonlinear and requires a certain initial guess. Fixed point iteration on the other hand appears insensitive to the initial guess.

YaqiWang commented:
@fdkong @friedmud This should be ready.

fdkong commented on framework/include/problems/FEProblemBase.h:
```diff
@@ -1157,6 +1157,7 @@ class FEProblemBase : public SubProblem, public Restartable
   virtual void cacheResidual(THREAD_ID tid) override;
   virtual void cacheResidualNeighbor(THREAD_ID tid) override;
   virtual void addCachedResidual(THREAD_ID tid) override;
+  virtual void addCachedResidualTags(THREAD_ID tid, const std::set<TagID> & tags) override;
```
I do not understand why we need this since low-level cached residuals are already handled in Assembly. `addCachedResidual` should be enough. If it is not working, we should fix this bug instead of creating a new unnecessary  API

```
void
Assembly::addCachedResiduals()
{
  for (auto tag = beginIndex(_cached_residual_values); tag < _cached_residual_values.size(); tag++)
  {
    if (!_sys.hasVector(tag))
    {
      _cached_residual_values[tag].clear();
      _cached_residual_rows[tag].clear();
      continue;
    }
    addCachedResidual(_sys.getVector(tag), tag);
  }
}
```

fdkong replied:
BTW, this also potentially makes the tags inconsistent.

YaqiWang replied:
This function only add residual to certain tagged vectors NOT all of them. That is the whole point. The nonlinear system computeResidual function is operating on certain tags, thus requires this function.

fdkong replied:
If you want to leave some tagged vectors alone, you can  just do `disassociateVectorFromTag`, and then corresponding vectors will be ignored.

YaqiWang replied:
Not in `NonlinearSystemBase::computeResidualInternal(const std::set<TagID> & tags)` without this fix. This is evidenced by the need of a `close` function call before this fix.

YaqiWang replied:
In the old way, you will basically touch all vectors no matter if they are disassociated or not.

fdkong replied:
We do not touch the disassociated vectors.

Again, you should not use `computeResidualInternal`

fdkong commented on framework/include/problems/SubProblem.h:
```diff
@@ -278,6 +278,7 @@ class SubProblem : public Problem
   virtual void cacheResidual(THREAD_ID tid) = 0;
   virtual void cacheResidualNeighbor(THREAD_ID tid) = 0;
   virtual void addCachedResidual(THREAD_ID tid) = 0;
+  virtual void addCachedResidualTags(THREAD_ID tid, const std::set<TagID> & tags) = 0;
```
Remove

fdkong commented on framework/src/problems/DisplacedProblem.C:
```diff
@@ -691,6 +691,13 @@ DisplacedProblem::addCachedResidual(THREAD_ID tid)
 }
 
+void
+DisplacedProblem::addCachedResidualTags(THREAD_ID tid, const std::set<TagID> & tags)
```
Similar

fdkong commented on framework/src/problems/FEProblemBase.C:
```diff
@@ -1183,6 +1183,16 @@ FEProblemBase::addCachedResidual(THREAD_ID tid)
     _displaced_problem->addCachedResidual(tid);
 }
 
+void
```
Similar

fdkong commented on framework/src/systems/NonlinearSystemBase.C:
```diff
@@ -1310,10 +1310,9 @@ NonlinearSystemBase::computeResidualInternal(const std::set<TagID> & tags)
-    for (unsigned int i = 0; i < n_threads;
-         i++) // Add any cached residuals that might be hanging around
-      _fe_problem.addCachedResidual(i);
+    // Add any cached residuals that might be hanging around
```
Similar

YaqiWang replied:
This is the API: `NonlinearSystemBase::computeResidualInternal(const std::set<TagID> & tags)`

fdkong replied:
You really should use `computeResidualTags` and do your own `associateVectorToTag` and `disassociateVectorFromTag`, and you can control the things whatever way you want.

fdkong commented on framework/src/systems/NonlinearSystemBase.C:
```diff
@@ -1380,10 +1379,9 @@ NonlinearSystemBase::computeResidualInternal(const std::set<TagID> & tags)
-        for (unsigned int i = 0; i < n_threads;
-             i++) // Add any cached residuals that might be hanging around
-          _fe_problem.addCachedResidual(i);
+        // Add any cached residuals that might be hanging around
```
Similar

fdkong commented on framework/src/systems/NonlinearSystemBase.C:
```diff
@@ -1402,10 +1400,9 @@ NonlinearSystemBase::computeResidualInternal(const std::set<TagID> & tags)
-      for (unsigned int i = 0; i < n_threads;
-           i++) // Add any cached residuals that might be hanging around
-        _fe_problem.addCachedResidual(i);
+      // Add any cached residuals that might be hanging around
```
Similar

fdkong commented:
Looks good for me. But I think we should NOT add the API `addCachedResidualTags`.

YaqiWang commented:
@fdkong I know what you mean. I think we possible have Solution 1 in #13189 in place first as a temporary solution. But here I will add disassociation in `FixedPointProblem::computeResidual` for now.

fdkong commented:
It looks pretty good for me. @friedmud  Do you some comments or concerns? If no comments, I am going to merge this.

friedmud reviewed (approved):
👍 

This looks good enough for now.  I think there could be a bit of refinement on the input file side before we put something like this in MOOSE.  But, it's a good first shot at something we've wanted for a while.
